src/Compiler/test/Microsoft.NET.Sdk.Razor.SourceGenerators.Tests/RazorSourceGeneratorTestsBase.cs

```diff
@@ -0,0 +1,332 @@
+        return (result.Item1, result.Item2);
+    }
+
+    protected static async ValueTask<(GeneratorDriver, ImmutableArray<AdditionalText>, TestAnalyzerConfigOptionsProvider)> GetDriverWithAdditionalTextAndProviderAsync(Project project, Action<TestAnalyzerConfigOptionsProvider>? configureGlobalOptions = null)
+    {
+        var razorSourceGenerator = new RazorSourceGenerator().AsSourceGenerator();
+        var driver = (GeneratorDriver)CSharpGeneratorDriver.Create(new[] { razorSourceGenerator }, parseOptions: (CSharpParseOptions)project.ParseOptions!, driverOptions: new GeneratorDriverOptions(IncrementalGeneratorOutputKind.None, true));
+
+        var optionsProvider = new TestAnalyzerConfigOptionsProvider();
+        optionsProvider.TestGlobalOptions["build_property.RazorConfiguration"] = "Default";
+        optionsProvider.TestGlobalOptions["build_property.RootNamespace"] = "MyApp";
+        optionsProvider.TestGlobalOptions["build_property.RazorLangVersion"] = "Latest";
+        optionsProvider.TestGlobalOptions["build_property.GenerateRazorMetadataSourceChecksumAttributes"] = "false";
+
+        var additionalTexts = ImmutableArray<AdditionalText>.Empty;
+
+        foreach (var document in project.AdditionalDocuments)
+        {
+            var additionalText = new TestAdditionalText(document.Name, await document.GetTextAsync());
+            additionalTexts = additionalTexts.Add(additionalText);
+
+            var additionalTextOptions = new TestAnalyzerConfigOptions
+            {
+                ["build_metadata.AdditionalFiles.TargetPath"] = Convert.ToBase64String(Encoding.UTF8.GetBytes(additionalText.Path)),
+            };
+
+            optionsProvider.AdditionalTextOptions[additionalText.Path] = additionalTextOptions;
+        }
+
+        configureGlobalOptions?.Invoke(optionsProvider);
+
+        driver = driver
+            .AddAdditionalTexts(additionalTexts)
+            .WithUpdatedAnalyzerConfigOptions(optionsProvider);
+
+        return (driver, additionalTexts, optionsProvider);
+    }
+
+    protected static GeneratorRunResult RunGenerator(Compilation compilation, ref GeneratorDriver driver, params Action<Diagnostic>[] expectedDiagnostics)
+    {
+        driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out compilation, out _);
+
+        var actualDiagnostics = compilation.GetDiagnostics().Where(d => d.Severity != DiagnosticSeverity.Hidden);
+        Assert.Collection(actualDiagnostics, expectedDiagnostics);
+
+        var result = driver.GetRunResult();
+        return result.Results.Single();
+    }
+
+    protected static Project CreateTestProject(
+        Dictionary<string, string> additonalSources,
+        Dictionary<string, string>? sources = null)
+    {
+        var project = CreateBaseProject();
+
+        if (sources is not null)
+        {
+            foreach (var (name, source) in sources)
```

Comment:
This is going to produce non-deterministic output because you're iterating over a dictionary. Should include an `.OrderBy(x => x.Key)` call here to get deterministic output.

Reply:
You are right, but that would require changing existing tests - where inputs and outputs could have switched order, e.g.,

```cs
var project = CreateTestProject(new()
{
    ["Pages/Index.razor"] = "<h1>Hello world</h1>",
    ["Pages/Counter.razor"] = "<h1>Counter</h1>",
});
// ...
result.VerifyPageOutput(
    "...Counter.razor output...",
    "...Index.razor output...");
```

(It actually behaves deterministically now even though it's using a dictionary.)

We ideally want to pass a normal ordered list of inputs. To avoid changing all the callers (but I guess that wouldn't be that bad either if you prefer), I wrote a simple `string`-typed `OrderedDictionary` wrapper.